transports/quic/src/transport.rs

```diff
@@ -0,0 +1,350 @@
+    <C::Session as Session>::PacketKey: Unpin,
+{
+    type Output = Result<(PeerId, QuicMuxer<C>), QuicError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        loop {
+            match &mut *self {
+                Self::Dialing(rx) => match Pin::new(rx).poll(cx) {
+                    Poll::Ready(Ok(Ok(muxer))) => {
+                        *self = Self::Upgrade(QuicUpgrade::new(muxer));
+                    }
+                    Poll::Ready(Ok(Err(err))) => return Poll::Ready(Err(err)),
+                    Poll::Ready(Err(_)) => panic!("endpoint crashed"),
+                    Poll::Pending => return Poll::Pending,
+                },
+                Self::Upgrade(upgrade) => return Pin::new(upgrade).poll(cx),
+            }
+        }
+    }
+}
+
+pub struct QuicUpgrade<C: Crypto> {
+    muxer: Option<QuicMuxer<C>>,
+}
+
+impl<C: Crypto> QuicUpgrade<C> {
+    fn new(muxer: QuicMuxer<C>) -> Self {
+        Self { muxer: Some(muxer) }
+    }
+}
+
+impl<C: Crypto> Future for QuicUpgrade<C>
+where
+    <C::Session as Session>::HeaderKey: Unpin,
+    <C::Session as Session>::PacketKey: Unpin,
+{
+    type Output = Result<(PeerId, QuicMuxer<C>), QuicError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let inner = Pin::into_inner(self);
+        let muxer = inner.muxer.as_mut().expect("future polled after ready");
+        match muxer.poll_event(cx) {
+            Poll::Pending => {
+                if let Some(peer_id) = muxer.peer_id() {
+                    muxer.set_accept_incoming(true);
+                    Poll::Ready(Ok((
+                        peer_id,
+                        inner.muxer.take().expect("future polled after ready"),
+                    )))
+                } else {
+                    Poll::Pending
+                }
+            }
+            Poll::Ready(Err(err)) => Poll::Ready(Err(err.into())),
+            Poll::Ready(Ok(_)) => {
+                panic!("muxer.incoming is set to false so no events can be produced");
+            }
+        }
+    }
+}
+
+/// Tries to turn a QUIC multiaddress into a UDP [`SocketAddr`]. Returns an error if the format
+/// of the multiaddr is wrong.
+fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Result<(SocketAddr, Option<PublicKey>), ()> {
+    let mut iter = addr.iter().peekable();
+    let proto1 = iter.next().ok_or(())?;
+    let proto2 = iter.next().ok_or(())?;
+    let proto3 = iter.next().ok_or(())?;
+
+    let peer_id = if let Some(Protocol::P2p(peer_id)) = iter.peek() {
+        if peer_id.code() != multihash::Code::Identity.into() {
+            return Err(());
+        }
+        let public_key =
+            libp2p_core::PublicKey::from_protobuf_encoding(peer_id.digest()).map_err(|_| ())?;
+        let public_key = if let libp2p_core::PublicKey::Ed25519(public_key) = public_key {
+            public_key.encode()
+        } else {
+            return Err(());
```

Comment:
This code does not support other peer ids than Ed25519

Reply:
fixed in https://github.com/dvc94ch/rust-libp2p/pull/13